### PR TITLE
[Windows] Shutdown the engine before destroying the surface

### PR DESCRIPTION
**Pull request by loic-sharma**
Currently, closing the window does the following:

1. The `FlutterWindowsView` [destroys its surface](https://github.com/flutter/engine/blob/df5e055f6aba457f8bee607c9dabccfcff15726c/shell/platform/windows/flutter_windows_view.cc#L52)
2. The `FlutterWindowsView` [destroys the `FlutterWindowsEngine` it owns](https://github.com/flutter/engine/blob/df5e055f6aba457f8bee607c9dabccfcff15726c/shell/platform/windows/flutter_windows_view.h#L349)
3. The `FlutterWindowsEngine` shuts down the real engine ([1](https://github.com/flutter/engine/blob/df5e055f6aba457f8bee607c9dabccfcff15726c/shell/platform/windows/flutter_windows_engine.cc#L234), [2](https://github.com/flutter/engine/blob/df5e055f6aba457f8bee607c9dabccfcff15726c/shell/platform/windows/flutter_windows_engine.cc#L410))

As a result, the raster thread may attempt to draw to a destroyed surface. This change ensures the engine is shutdown before destroying the view's resources. This ensures the raster thread is stopped before the underlying surface is destroyed.

Addresses https://github.com/flutter/flutter/issues/124463

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/windows/flutter_windows_view_unittests.cc [hunk 1]
```diff
@@ -94,15 +103,57 @@ std::unique_ptr<FlutterWindowsEngine> GetTestEngine() {
...
+      std::make_unique<NiceMock<MockWindowBindingHandler>>();
+  std::unique_ptr<MockAngleSurfaceManager> surface_manager =
+      std::make_unique<MockAngleSurfaceManager>();
+
+  EngineModifier modifier(engine.get());
+  FlutterWindowsView view(std::move(window_binding_handler));
+
+  // The engine must be stopped before the surface can be destroyed.
+  InSequence s;
+  EXPECT_CALL(*engine.get(), Stop);
```
Comment (yaakovschectman) [inline shell/platform/windows/flutter_windows_view_unittests.cc:143]: Minor nit: Should this specify a `Times`?

**Review by yaakovschectman — APPROVED** [review]
LGTM save for one nit. Thanks!

File: shell/platform/windows/flutter_windows_view_unittests.cc (hunk 1 above)
Comment (loic-sharma) [reply shell/platform/windows/flutter_windows_view_unittests.cc]: Good idea, updated!

**Review by cbracken — APPROVED** [review]
Looks good.

I can't remember, do we allow headless execution on Windows? If so, we probably want to only shutdown if headless execution isn't permitted. On macOS/iOS we do this by setting an ivar in our engine wrapper. On Linux we have a separate headless renderer.

**Comment by loic-sharma** [comment]
> I can't remember, do we allow headless execution on Windows?

Kind of. On Windows you can create an engine, run it, and then later attach a view controller. During that period, Windows is effectively running in headless mode. In this headless mode, the Windows embedder will ignore attempts to render. **However**, once a view controller has been attached the Windows engine cannot transition back to headless mode. The view controller owns the engine, destroying the view controller will shutdown the engine. This is something we'll fix as part of the multi-view work :)

**Pull request merged**